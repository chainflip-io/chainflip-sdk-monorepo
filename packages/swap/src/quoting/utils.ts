import assert from 'assert';
import BigNumber from 'bignumber.js';
import { filter, Observable, Subscription } from 'rxjs';
import { Side, type SwapInput } from '@/amm-addon';
import { InternalAsset, getAssetAndChain } from '@/shared/enums';
import { MarketMakerQuote, Leg as MarketMakerLeg } from './schemas';
import env from '../config/env';
import { getAssetPrice } from '../pricing';
import { assertUnreachable } from '../utils/function';
import { SwapRateArgs } from '../utils/statechain';

export type QuoteType = 'pool' | 'market_maker';

export const collectMakerQuotes = (
  requestId: string,
  expectedQuotes: number,
  quotes$: Observable<{ marketMaker: string; quote: MarketMakerQuote }>,
): Promise<MarketMakerQuote[]> => {
  if (expectedQuotes === 0) return Promise.resolve([]);

  const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();

  return new Promise((resolve) => {
    let sub: Subscription;

    const complete = () => {
      sub.unsubscribe();
      resolve([...clientsReceivedQuotes.values()]);
    };

    sub = quotes$
      .pipe(filter(({ quote }) => quote.request_id === requestId))
      .subscribe(({ marketMaker, quote }) => {
        clientsReceivedQuotes.set(marketMaker, quote);
        if (clientsReceivedQuotes.size === expectedQuotes) complete();
      });

    setTimeout(complete, env.QUOTE_TIMEOUT);
  });
};

export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
  const price = await getAssetPrice(asset);

  if (typeof price !== 'number') return null;

  return BigInt(new BigNumber(amount).times(price).toFixed(0));
};

type BaseAsset = Exclude<InternalAsset, 'Usdc'>;

export class Leg {
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint): Leg;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null): Leg | null;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null) {
    if (amount === null) return null;
    assert(baseAsset !== quoteAsset, 'baseAsset and quoteAsset must be different');
    assert(baseAsset === 'Usdc' || quoteAsset === 'Usdc', 'one of the assets must be USDC');
    return new Leg(baseAsset, quoteAsset, amount);
  }

  private constructor(
    private readonly baseAsset: InternalAsset,
    private readonly quoteAsset: InternalAsset,
    public amount: bigint,
  ) {}

  toPoolJSON(): SwapRateArgs {
    return {
      amount: this.amount,
      srcAsset: this.baseAsset,
      destAsset: this.quoteAsset,
    };
  }

  private getSide(): 'BUY' | 'SELL' {
    return this.quoteAsset !== 'Usdc' ? 'BUY' : 'SELL';
  }

  toMarketMakerJSON(): MarketMakerLeg {
    const side = this.getSide();
    let normalizedBaseAsset: BaseAsset;
    const normalizedQuoteAsset = 'Usdc';

    if (this.quoteAsset !== 'Usdc') {
      normalizedBaseAsset = this.quoteAsset;
    } else if (this.baseAsset !== 'Usdc') {
      normalizedBaseAsset = this.baseAsset;
    } else {
      return assertUnreachable('invalid leg');
    }

    return {
      base_asset: getAssetAndChain(normalizedBaseAsset),
      quote_asset: getAssetAndChain(normalizedQuoteAsset),
      amount: this.amount.toString(),
      side,
    };
  }

  toSwapInput(): SwapInput {
    return {
      side: this.getSide() === 'BUY' ? Side.Buy : Side.Sell,
      amount: this.amount,
      limitOrders: [],
    };
  }
}
