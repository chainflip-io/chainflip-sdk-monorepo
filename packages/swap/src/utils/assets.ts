import { SupportedAsset, supportedAsset } from '@/shared/enums';

export const isSupportedAsset = (value: string): value is SupportedAsset =>
  supportedAsset.safeParse(value).success;

export function assertSupportedAsset(
  value: string,
): asserts value is SupportedAsset {
  if (!isSupportedAsset(value)) {
    throw new Error(`received invalid asset "${value}"`);
  }
}

export const decimalPlaces: Record<SupportedAsset, number> = {
  DOT: 10,
  ETH: 18,
  FLIP: 18,
  USDC: 6,
  BTC: 8,
};
