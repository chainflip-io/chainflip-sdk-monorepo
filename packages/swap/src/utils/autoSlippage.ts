import BigNumber from 'bignumber.js';
import { assetConstants, chainConstants } from '@/shared/enums';
import { getDeployedLiquidity, getUndeployedLiquidity } from './pools';
import { getRequiredBlockConfirmations } from './rpc';
import { InternalAsset } from '../client';

const getLiquidityAutoSlippage = async (
  srcAsset: InternalAsset,
  destAsset: InternalAsset,
  amount: bigint,
) => {
  const deployedLiquidity = await getDeployedLiquidity(srcAsset, destAsset);
  const liquidityRatio = new BigNumber(amount.toString()).div(deployedLiquidity.toString());

  if (liquidityRatio.isGreaterThan(0.2)) {
    return liquidityRatio.multipliedBy(2);
  }
  return new BigNumber(0);
};

const getDepositTimeAutoSlippage = async (srcAsset: InternalAsset, isBoosted: boolean) => {
  const { chain } = assetConstants[srcAsset];
  const { blockTimeSeconds } = chainConstants[chain];
  const blockConfirmations = isBoosted ? 1 : (await getRequiredBlockConfirmations(srcAsset)) ?? 0;
  const depositTimeMinutes = (blockConfirmations * blockTimeSeconds) / 60;

  if (['Solana', 'Arbitrum'].includes(chain)) {
    return depositTimeMinutes * 0.01;
  }
  if (['Ethereum', 'Polkadot'].includes(chain)) {
    return depositTimeMinutes * 0.05;
  }
  return depositTimeMinutes * 0.1;
};

const getUndeployedLiquidityAutoSlippage = async (asset: InternalAsset, amount: bigint) => {
  const undeployedLiquidity = await getUndeployedLiquidity(asset);
  if (undeployedLiquidity >= amount) {
    return new BigNumber(-0.5);
  }

  const undeployedRatio = new BigNumber(undeployedLiquidity.toString()).div(amount.toString());
  if (undeployedRatio.gt(0.5)) {
    return new BigNumber(-undeployedRatio * 0.1);
  }

  return new BigNumber(0);
};

export const calculateRecommendedSlippage = async ({
  srcAsset,
  destAsset,
  intermediateAmount,
  egressAmount,
  boostFeeBps,
  dcaChunks,
}: {
  srcAsset: InternalAsset;
  destAsset: InternalAsset;
  intermediateAmount?: bigint;
  egressAmount: bigint;
  boostFeeBps?: number;
  dcaChunks: number;
}) => {
  const minSlippage = 0.1;
  const baseSlippage = 1;

  const assetTo = srcAsset === 'Usdc' || destAsset === 'Usdc' ? destAsset : 'Usdc';

  const [deployedLiquiditySlippage, timeSlippage, undeployedLiquiditySlippage] = await Promise.all([
    await getLiquidityAutoSlippage(
      srcAsset,
      assetTo,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
    await getDepositTimeAutoSlippage(srcAsset, Boolean(boostFeeBps && boostFeeBps > 0)),
    await getUndeployedLiquidityAutoSlippage(
      intermediateAmount ? 'Usdc' : destAsset,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
  ]);

  const calculatedSlippage = deployedLiquiditySlippage
    .plus(timeSlippage)
    .plus(undeployedLiquiditySlippage)
    .plus(baseSlippage);

  return Math.max(minSlippage, Math.round(calculatedSlippage.toNumber() * 100) / 100);
};
