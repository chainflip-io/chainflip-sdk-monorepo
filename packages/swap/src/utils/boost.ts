import { ONE_IN_PIP, bigintMin, getPipAmountFromAmount } from '@/shared/functions';
import { BoostPoolsDepth } from '@/shared/rpc';

export const getBoostFeeBpsForAmount = async ({
  amount,
  assetBoostPoolsDepth,
}: {
  amount: bigint;
  assetBoostPoolsDepth: BoostPoolsDepth;
}): Promise<number | undefined> => {
  assetBoostPoolsDepth.sort((a, b) => (a.tier < b.tier ? -1 : 1));

  let _amount = amount;
  let feeAmount = 0n;

  for (const poolDepth of assetBoostPoolsDepth) {
    const poolAvailableAmount = poolDepth.availableAmount;

    const amountToBeUsedFromPool = bigintMin(_amount, poolAvailableAmount);
    feeAmount += getPipAmountFromAmount(amountToBeUsedFromPool, poolDepth.tier);
    _amount -= amountToBeUsedFromPool;
  }

  // Not enough liquidity in the boost pools
  if (_amount > 0) return undefined;

  return Number((feeAmount * BigInt(ONE_IN_PIP)) / BigInt(amount));
};
