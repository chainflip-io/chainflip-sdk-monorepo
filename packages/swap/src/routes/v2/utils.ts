import { getAssetAndChain } from '@/shared/enums';
import { assertUnreachable } from '@/shared/functions';
import { StateV2 } from './swap';
import prisma, {
  Prisma,
  Swap,
  Egress,
  Broadcast,
  SwapFee,
  IgnoredEgress,
  FailedSwap,
} from '../../client';
import { getPendingBroadcast } from '../../ingress-egress-tracking';
import { failedSwapMessage } from '../../utils/swap';
import { FailureMode } from '../swap';

export const depositChannelInclude = {
  failedBoosts: true,
  failedSwaps: true,
  affiliates: {
    select: {
      account: true,
      commissionBps: true,
    },
  },
} as const;

export const swapRequestInclude = {
  swaps: { include: { fees: true }, orderBy: { nativeId: 'desc' } },
  egress: { include: { broadcast: true } },
  refundEgress: { include: { broadcast: true } },
  fees: true,
  ignoredEgresses: true,
  swapDepositChannel: {
    include: depositChannelInclude,
  },
} as const;

export const getSwapFields = (swap: Swap & { fees: SwapFee[] }) => ({
  swapInputAmount: swap.swapInputAmount.toString(),
  swapOutputAmount: swap.swapOutputAmount?.toString(),
  scheduledAt: swap.swapScheduledAt.valueOf(),
  scheduledBlockIndex: swap.swapScheduledBlockIndex ?? undefined,
  executedAt: swap.swapExecutedAt?.valueOf(),
  executedBlockIndex: swap.swapExecutedBlockIndex ?? undefined,
  retryCount: swap.retryCount,
  latestSwapRescheduledAt: swap.latestSwapRescheduledAt?.valueOf(),
  latestSwapRescheduledBlockIndex: swap.latestSwapRescheduledBlockIndex ?? undefined,
  fees: swap.fees.map((fee) => ({
    type: fee.type,
    ...getAssetAndChain(fee.asset),
    amount: fee.amount.toString(),
  })),
});

export const getDepositIgnoredFailedState = (failedSwap: FailedSwap) => ({
  failedAt: failedSwap.failedAt.valueOf(),
  failedAtBlockIndex: failedSwap.failedBlockIndex,
  mode: FailureMode.IngressIgnored,
  reason: {
    name: failedSwap.reason,
    message: failedSwapMessage[failedSwap.reason],
  },
});

export const getEgressFailureState = async (
  ignoredEgress: IgnoredEgress | undefined,
  broadcast: Broadcast | null | undefined,
  type: IgnoredEgress['type'] | undefined,
) => {
  const abortedBroadcast = broadcast?.abortedAt?.valueOf();
  const failedAt = abortedBroadcast ?? ignoredEgress?.ignoredAt?.valueOf();
  const failedAtBlockIndex = broadcast?.abortedBlockIndex ?? ignoredEgress?.ignoredBlockIndex;

  let error;
  let failureMode;

  if (failedAt && failedAtBlockIndex) {
    error = {
      name: 'Unknown',
      message: 'An unknown error occurred',
    };
    if (ignoredEgress) {
      switch (ignoredEgress.type) {
        case 'REFUND':
          failureMode = FailureMode.RefundEgressIgnored;
          break;
        case 'SWAP':
          failureMode = FailureMode.EgressIgnored;
          break;
        default:
          assertUnreachable(ignoredEgress.type);
      }
      const [stateChainError] = await Promise.all([
        prisma.stateChainError.findUniqueOrThrow({
          where: { id: ignoredEgress.stateChainErrorId },
        }),
      ]);
      error = {
        name: stateChainError.name,
        message: stateChainError.docs,
      };
    } else if (abortedBroadcast) {
      const message =
        type === 'SWAP' ? 'The swap broadcast was aborted' : 'The refund broadcast was aborted';
      failureMode = FailureMode.BroadcastAborted;
      error = {
        name: 'BroadcastAborted',
        message,
      };
    }
    return {
      failedAt,
      failedAtBlockIndex,
      mode: failureMode,
      reason: error,
    };
  }
  return undefined;
};

export const getEgressStatusFields = async (
  egress: Egress | null | undefined,
  broadcast: Broadcast | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  type: IgnoredEgress['type'] | undefined,
  egressTrackerTxRef: string | null | undefined,
) => {
  const ignoredEgress = ignoredEgresses?.find((e) => e.type === type);
  const failureState = await getEgressFailureState(ignoredEgress, broadcast, type);
  return {
    ...(egress && {
      outputAmount: egress.amount?.toFixed(),
      scheduledAt: egress.scheduledAt?.valueOf(),
      scheduledBlockIndex: egress.scheduledBlockIndex ?? undefined,
    }),
    ...(broadcast && {
      sentAt: broadcast?.succeededAt?.valueOf(),
      sentAtBlockIndex: broadcast?.succeededBlockIndex ?? undefined,
      sentTxRef: broadcast?.transactionRef ?? egressTrackerTxRef,
    }),
    ...(failureState && { failure: failureState }),
    ...(ignoredEgress && { ignoredAmount: ignoredEgress.amount?.toFixed() }),
  };
};

export const getSwapState = async (
  failedSwap: FailedSwap | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  swapRequest:
    | Prisma.SwapRequestGetPayload<{
        include: typeof swapRequestInclude;
      }>
    | undefined
    | null,
) => {
  let state: StateV2 | undefined;
  let swapEgressTrackerTxRef: string | null | undefined;
  let refundEgressTrackerTxRef: string | null | undefined;
  const swapEgress = swapRequest?.egress;
  const refundEgress = swapRequest?.refundEgress;
  const egress = swapEgress ?? refundEgress;

  if (
    failedSwap ||
    (ignoredEgresses && ignoredEgresses.length > 0) ||
    egress?.broadcast?.abortedAt
  ) {
    state = StateV2.Failed;
  } else if (egress?.broadcast?.succeededAt) {
    state = StateV2.Completed;
  } else if (egress?.broadcast) {
    if (swapEgress?.broadcast) {
      state = StateV2.Sending;
      const pendingSwapBroadcast = await getPendingBroadcast(swapEgress.broadcast);
      if (pendingSwapBroadcast) {
        state = StateV2.Sent;
        swapEgressTrackerTxRef = pendingSwapBroadcast.tx_ref;
      }
    }
    if (refundEgress?.broadcast) {
      const pendingRefundBroadcast = await getPendingBroadcast(refundEgress.broadcast);
      if (pendingRefundBroadcast) {
        state = StateV2.Sent;
        refundEgressTrackerTxRef = pendingRefundBroadcast.tx_ref;
      }
    }
  } else if (egress) {
    state = StateV2.Sending;
  } else if (swapRequest?.swaps.some((s) => s.swapScheduledAt)) {
    state = StateV2.Swapping;
  } else {
    state = StateV2.Receiving;
  }

  return {
    state,
    swapEgressTrackerTxRef,
    refundEgressTrackerTxRef,
  };
};
