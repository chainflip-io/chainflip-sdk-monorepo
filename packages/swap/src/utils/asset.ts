import { ChainflipAsset } from '@chainflip/utils/chainflip';

export const isAssetHub = (asset: ChainflipAsset): asset is 'HubDot' | 'HubUsdc' | 'HubUsdt' =>
  asset === 'HubDot' || asset === 'HubUsdc' || asset === 'HubUsdt';
