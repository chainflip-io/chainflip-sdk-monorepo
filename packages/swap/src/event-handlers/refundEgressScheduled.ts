import { swappingRefundEgressScheduled as schema150 } from '@chainflip/processor/150/swapping/refundEgressScheduled';
import { swappingRefundEgressScheduled as schema160 } from '@chainflip/processor/160/swapping/refundEgressScheduled';
import { z } from 'zod';
import type { EventHandlerArgs } from '.';

const eventArgs = z.union([
  schema160,
  schema150.transform(({ swapId, ...rest }) => ({
    swapRequestId: swapId,
    ...rest,
  })),
]);

export default async function refundEgressScheduled({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const {
    swapRequestId,
    egressId: [chain, nativeId],
    egressFee,
    amount,
    asset,
  } = eventArgs.parse(event.args);

  await prisma.swapRequest.update({
    where: { nativeId: swapRequestId },
    data: {
      refundEgress: {
        create: {
          nativeId,
          chain,
          amount: amount.toString(),
          scheduledAt: new Date(block.timestamp),
          scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
        },
      },
      fees: {
        create: {
          type: 'EGRESS',
          asset,
          amount: egressFee.toString(),
        },
      },
    },
  });
}
