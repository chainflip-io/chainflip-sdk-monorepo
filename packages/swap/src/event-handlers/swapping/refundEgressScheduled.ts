import { swappingRefundEgressScheduled as schema160 } from '@chainflip/processor/160/swapping/refundEgressScheduled';
import { swappingRefundEgressScheduled as schema180 } from '@chainflip/processor/180/swapping/refundEgressScheduled';
import z from 'zod';
import type { EventHandlerArgs } from '../index';

const eventArgs = z.union([
  schema180,
  schema160.transform((data) => ({
    ...data,
    egressFee: [data.egressFee, data.asset] as const,
  })),
]);

export default async function refundEgressScheduled({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const {
    swapRequestId,
    egressId: [chain, nativeId],
    egressFee: [egressFee, egressFeeAsset],
    amount,
  } = eventArgs.parse(event.args);

  await prisma.swapRequest.update({
    where: { nativeId: swapRequestId },
    data: {
      refundEgress: {
        create: {
          nativeId,
          chain,
          amount: amount.toString(),
          scheduledAt: new Date(block.timestamp),
          scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
        },
      },
      fees: {
        create: {
          type: 'EGRESS',
          asset: egressFeeAsset,
          amount: egressFee.toString(),
        },
      },
    },
  });
}
