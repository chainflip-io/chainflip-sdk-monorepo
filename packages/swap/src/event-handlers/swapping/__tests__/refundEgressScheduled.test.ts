import { describe, it, expect, beforeEach } from 'vitest';
import prisma from '../../../client.js';
import { check, DOT_ADDRESS, refundEgressScheduledMock } from '../../__tests__/utils.js';
import refundEgressScheduled, { RefundEgressScheduledArgs } from '../refundEgressScheduled.js';

describe(refundEgressScheduled, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap", "Egress", "SwapRequest" CASCADE`;
  });

  it('creates egress for refund on an existing swap', async () => {
    const { event, block } = refundEgressScheduledMock;
    const swapId = BigInt(event.args.swapRequestId);

    await prisma.swapRequest.create({
      data: {
        nativeId: swapId,
        depositAmount: '10000000000',
        swapInputAmount: '10000000000',
        depositFinalisedAt: new Date(block.timestamp - 12000),
        depositFinalisedBlockIndex: `${block.height - 100}-${event.indexInBlock}`,
        srcAsset: 'Eth',
        destAsset: 'Dot',
        destAddress: DOT_ADDRESS,
        requestType: 'LEGACY_SWAP',
        originType: 'VAULT',
        swapRequestedAt: new Date(block.timestamp - 12000),
        swapRequestedBlockIndex: '92-398',
        totalBrokerCommissionBps: 0,
      },
    });

    await refundEgressScheduled({ block, event, prisma });

    const swapRequest = await prisma.swapRequest.findUniqueOrThrow({
      where: { nativeId: swapId },
      include: {
        refundEgress: true,
        fees: { select: { amount: true, asset: true, type: true } },
      },
    });

    expect(swapRequest).toMatchSnapshot({
      id: expect.any(BigInt),
      refundEgressId: expect.any(BigInt),
      refundEgress: {
        id: expect.any(BigInt),
      },
    });
  });
  it('calculates the refund fee', async () => {
    const { event, block } = refundEgressScheduledMock;
    const swapId = BigInt(event.args.swapRequestId);

    await prisma.swapRequest.create({
      data: {
        nativeId: swapId,
        depositAmount: '10000000000',
        swapInputAmount: '9500000000',
        depositFinalisedAt: new Date(block.timestamp - 12000),
        depositFinalisedBlockIndex: `${block.height - 100}-${event.indexInBlock}`,
        srcAsset: 'Eth',
        destAsset: 'Dot',
        destAddress: DOT_ADDRESS,
        requestType: 'LEGACY_SWAP',
        originType: 'VAULT',
        swapRequestedAt: new Date(block.timestamp - 12000),
        swapRequestedBlockIndex: '92-398',
        totalBrokerCommissionBps: 0,
        swaps: {
          create: {
            srcAsset: 'Eth',
            destAsset: 'Dot',
            swapScheduledAt: new Date(block.timestamp - 30000),
            swapScheduledBlockIndex: `${block.height - 90}-${event.indexInBlock}`,
            nativeId: BigInt(1),
            type: 'SWAP',
            swapExecutedAt: new Date(block.timestamp - 12000),
            swapExecutedBlockIndex: `${block.height - 100}-${event.indexInBlock}`,
            swapInputAmount: '9000000000',
            swapOutputAmount: '10000000000',
          },
        },
      },
    });

    await refundEgressScheduled({
      block,
      event: {
        ...event,
        args: check<RefundEgressScheduledArgs>({
          swapRequestId: '9876545',
          egressFee: ['1000', { __kind: 'Eth' }],
          refundFee: '0x75bcd15',
          asset: { __kind: 'Eth' },
          amount: '400000000',
          egressId: [{ __kind: 'Ethereum' }, '1'] as const,
        }),
      },
      prisma,
    });

    const swapRequest = await prisma.swapRequest.findUniqueOrThrow({
      where: { nativeId: swapId },
      include: {
        refundEgress: true,
        fees: { select: { amount: true, asset: true, type: true } },
      },
    });

    expect(swapRequest).toMatchSnapshot({
      id: expect.any(BigInt),
      refundEgressId: expect.any(BigInt),
      refundEgress: {
        id: expect.any(BigInt),
      },
    });
  });
});
