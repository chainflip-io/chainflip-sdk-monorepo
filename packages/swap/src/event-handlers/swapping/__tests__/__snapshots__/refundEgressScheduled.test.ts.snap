// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`refundEgressScheduled > calculates the refund fee 1`] = `
{
  "ccmGasBudget": null,
  "ccmMessage": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "dcaChunkIntervalBlocks": null,
  "dcaNumberOfChunks": null,
  "depositAmount": "10000000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2022-12-06T14:31:33.000Z,
  "depositFinalisedBlockIndex": "20-0",
  "depositTransactionRef": null,
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "Dot",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fallbackEgressId": null,
  "fees": [
    {
      "amount": "1000",
      "asset": "Eth",
      "type": "EGRESS",
    },
    {
      "amount": "123456789",
      "asset": "Eth",
      "type": "REFUND",
    },
  ],
  "fokMinPriceX128": null,
  "fokRefundAddress": null,
  "fokRetryDurationBlocks": null,
  "id": Any<BigInt>,
  "maxBoostFeeBps": 0,
  "nativeId": 9876545n,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgress": {
    "amount": "400000000",
    "broadcastId": null,
    "chain": "Ethereum",
    "fallbackDestinationAddress": null,
    "id": Any<BigInt>,
    "nativeId": 1n,
    "scheduledAt": 2022-12-06T14:31:45.000Z,
    "scheduledBlockIndex": "120-0",
  },
  "refundEgressId": Any<BigInt>,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapInputAmount": "9500000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2022-12-06T14:31:33.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;

exports[`refundEgressScheduled > calculates the refund fee for schema 190 1`] = `
{
  "ccmGasBudget": null,
  "ccmMessage": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "dcaChunkIntervalBlocks": null,
  "dcaNumberOfChunks": null,
  "depositAmount": "10000000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2022-12-06T14:31:33.000Z,
  "depositFinalisedBlockIndex": "20-0",
  "depositTransactionRef": null,
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "Dot",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fallbackEgressId": null,
  "fees": [
    {
      "amount": "1000",
      "asset": "Eth",
      "type": "EGRESS",
    },
    {
      "amount": "99999000",
      "asset": "Eth",
      "type": "REFUND",
    },
  ],
  "fokMinPriceX128": null,
  "fokRefundAddress": null,
  "fokRetryDurationBlocks": null,
  "id": Any<BigInt>,
  "maxBoostFeeBps": 0,
  "nativeId": 9876545n,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgress": {
    "amount": "400000000",
    "broadcastId": null,
    "chain": "Ethereum",
    "fallbackDestinationAddress": null,
    "id": Any<BigInt>,
    "nativeId": 1n,
    "scheduledAt": 2022-12-06T14:31:45.000Z,
    "scheduledBlockIndex": "120-0",
  },
  "refundEgressId": Any<BigInt>,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapInputAmount": "9500000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2022-12-06T14:31:33.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;

exports[`refundEgressScheduled > creates egress for refund on an existing swap 1`] = `
{
  "ccmGasBudget": null,
  "ccmMessage": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "dcaChunkIntervalBlocks": null,
  "dcaNumberOfChunks": null,
  "depositAmount": "10000000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2022-12-06T14:31:33.000Z,
  "depositFinalisedBlockIndex": "20-0",
  "depositTransactionRef": null,
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "Dot",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fallbackEgressId": null,
  "fees": [
    {
      "amount": "1000000000",
      "asset": "Eth",
      "type": "EGRESS",
    },
  ],
  "fokMinPriceX128": null,
  "fokRefundAddress": null,
  "fokRetryDurationBlocks": null,
  "id": Any<BigInt>,
  "maxBoostFeeBps": 0,
  "nativeId": 9876545n,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgress": {
    "amount": "10000000000",
    "broadcastId": null,
    "chain": "Ethereum",
    "fallbackDestinationAddress": null,
    "id": Any<BigInt>,
    "nativeId": 1n,
    "scheduledAt": 2022-12-06T14:31:45.000Z,
    "scheduledBlockIndex": "120-0",
  },
  "refundEgressId": Any<BigInt>,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapInputAmount": "10000000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2022-12-06T14:31:33.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;
